### ISSUE TITLE
Cuda error when using analog_tile.get_weights

### ISSUE THREAD

**Issue by nkyungmi** [IBM/aihwkit#163]
## Description

Kernel stops when using analog_tile.get_weights while working on cuda

## How to reproduce

1. Create model which works on cuda (by using `model.cuda()`)

 [`rpu = UnitCellRPUConfig(device=LinearStepDevice()` and `rpu = UnitCellRPUConfig(device=SoftBoundDevice()`, both end up with error]

[`model = AnalogSequential(fc1,relu,fc2,relu,fc3)` is used, and each fc1,fc2,fc3 is created from `AnalogLinear`]

2. Try `print(model[0].analog_tile.get_weights(realistic=False))`

3. Kernel Stops
## Expected behavior

Oddly enough, it **works fine in cpu environment** as this image.

[![aa.png](https://i.postimg.cc/9fFgtTzz/aa.png)](https://postimg.cc/bZWRqZxj)

## Other information

* Pytorch version: 1.7.1
* Package version: git clone
* OS: Ubuntu 18.04.2
* Python version: 3.9.1
* Conda version (or N/A):
* NVIDIA Graphic Driver: 450.51.05
* CUDA Version: 11.0
* GPU: RTX 6000

I'm using the same server with @chaeunl, and I think #113 is related to this problem.

**Comment by diego-plan9** [comment]
Thanks for reporting - unfortunately, as with #113, I still cannot reproduce the issue. Can you paste your full snippet? I am testing with:

```python
from torch.nn import ReLU

from aihwkit.nn import AnalogLinear, AnalogSequential
from aihwkit.simulator.configs import UnitCellRPUConfig
from aihwkit.simulator.configs.devices import LinearStepDevice, SoftBoundsDevice

# rpu = UnitCellRPUConfig(device=LinearStepDevice())
rpu = UnitCellRPUConfig(device=SoftBoundsDevice())

model = AnalogSequential(
    AnalogLinear(4, 2, rpu_config=rpu),
    ReLU(),
    AnalogLinear(2, 2, rpu_config=rpu),
    ReLU(),
    AnalogLinear(2, 1, rpu_config=rpu)
)

model = model.cuda()

print(model[0].analog_tile.get_weights(realistic=False))
```

```
(tensor([[ 0.3050, -0.2856, -0.2256,  0.4839],
        [-0.1653,  0.4160, -0.0682,  0.3711]], device=‘cuda:0’), tensor([-0.0620, -0.2205], device=‘cuda:0’))
```

It might be an specific combination of versions - on this run, I used `master` with `python 3.9`, `torch 1.7.1` and `cuda 11.2`. 

As a small aside:
```
UnitCellRPUConfig(device=SoftBoundsDevice())
```
is not fully correct (if using a `UnitCellRPUConfig`, the `device` should be a `UnitCell` - and if using a `SoftBoundsDevice` device, the config should be `SingleRPUConfig`). This should not cause the segmentation fault, though!

**Comment by nkyungmi** [comment]
First of all, thank you for your answers!!
My snippet is just same as what you'd written down, so it seems like it's a problem of my program version settings.
I'm trying to follow your combination of versions, but how did you set `cuda 11.2` on` torch 1.7.1`? 
Official website shows the way of installing pytorch with `cuda 11.1` & `cuda 10.2`..!

**Comment by chaeunl** [comment]
@diego-plan9 , thank you for your description. We don't know what the problem is, but as of now, there are no issues on this problem. NVIDIA driver=455.23.05, CUDA ver.=11.1, and PyTorch ver.=1.8.1.

I think there are some issues on PyTorch and CUDA compatibility because when installing PyTorch, CUDA bin file is also downloaded.  So, I guess that aihwkit and PyTorch are entangled with different CUDA version. It might be useful for potential users to let them give the compatibility table. 

Thank you for your help so far.

**Comment by diego-plan9** [comment]
Thanks for the update @chaeunl, and glad to see that the combination you mention seems to be working!

**Comment by diego-plan9** [comment]
Closing the issue as a new release has been made `0.4.0` (which includes fixes such as #250) and other versions have moved forward - please submit a new issue if the problem still persists, so we can track it accordingly with a clean state.
